### DISCUSSION TITLE
How to ensure different distractor images are shown an equal number of times?

### DISCUSSION THREAD

**Discussion by ellamoeck** [jspsych/jsPsych#777] [Q&A] [answered]
Hi everyone,
I am very new to jspsych and am trying to quickly adapt an existing script for an honours student to collect data online (due to COVID). It's an image-based RSVP program with fillers, distractors, and targets. The fillers, distractors, and targets are randomly sampled from image arrays. This random sampling is great for the fillers and targets (where it doesn't matter if participants see the same image multiple times), but I need to ensure that each distractor image is shown an equal number of times. I think I need to change something from the following code which appears at the end of the script, in which the shuffleArray is what's used for the filler and targets, and the sampleArray for the distractors.
Any help would be greatly appreciated!
Ella

```javascript
function shuffleArray(myArray) {
    var randNum, tempStore, j;
    for (j = myArray.length; j; j--) {
        randNum = Math.floor(Math.random() * j);
        tempStore = myArray[j - 1];
        myArray[j - 1] = myArray[randNum];
        myArray[randNum] = tempStore;
    }
};

function sampleArray(myArray) {
    return myArray[Math.floor(Math.random() * myArray.length)];
};
```

**Comment by vijaymarupudi** [comment, marked as answer]
How are these functions used in the current script?

**Reply by ellamoeck** [reply]
They are used in the rsvp trial sequence, which I've pasted below. The rewFilenames and neutFilenames are the distractors and StandardFilenames are the fillers. I can send through the full script, if that's helpful?

```javascript
shuffleArray(standardFilenames);

stimArray = [];
for (ii = 0; ii < itemsPerStream + 2; ii++) {
  // Extra 2 elements in array hold distractor and target stimuli; these will be moved into position later
  stimArray[ii] = standardFilenames[ii];
}

trialDistractType = distractType[trialSelector[trialNumInBlock]];

if (trialDistractType == 1) {
  distractID = sampleArray(rewFilenames);
} else if (trialDistractType == 2) {
  distractID = sampleArray(neutFilenames);
} else if (trialDistractType == 3) {
  distractID = standardFilenames[itemsPerStream];
}

stimArray[itemsPerStream] = distractID;

trialTargetType = Math.floor(Math.random() * 2);
if (trialTargetType == 0) {
  targetID = sampleArray(target0Filenames);
} else {
  targetID = sampleArray(target1Filenames);
}

stimArray[itemsPerStream + 1] = targetID;

return stimArray;
```

**Reply by vijaymarupudi** [reply]
Full script will be helpful for context. Could you explain your task in more detail too? I'm having trouble understanding it (I lack the background knowledge!)

**Reply by ellamoeck** [reply]
For sure. RSVP stands for Rapid Serial Visual Presentation. It's where a stream of images are briefly presented (in this case, it's a stream of 16 images presented for 100ms each). Most of the images in the stream are filler images (pictures of nature/buildings in this experiment), and one of them is the target (a nature/building image rotated 90 degrees to the left or the right). The task is to report which way the target image is rotated. But on 80% of trials, there's also a distractor image which is from one of four categories (though because of what this script was previously used for there's just two categories, "reward" or "neutral", in the attached). Ideally, I would have the four categories of distractors listed in the code, but when I've been trying to do that the code no longer runs. However, by getting each image to show an equal number of times, I thought I could get around the number of categories issue.

Note: There are a few things in the code which I know I don't need (like counterbalancing and some of the instructions), but I kept finding that when I deleted them the code would no longer run. My primary interest is in useable, rather than beautiful, code given the time pressure on getting this study up and running.

Many thanks for your help.

[RSVPTask.html.zip](https://github.com/jspsych/jsPsych/files/4822276/RSVPTask.html.zip)

**Comment by vijaymarupudi** [comment]
So the participant sees a stream of images. When they see a filler, they don't respond. When they see a distractor, they don't respond, but when they see a target, they respond with either left or right, correct?

The actual position of the filler items doesn't not matter, nor do the targets, and both filler items and targets can be repeated. However, you have a specific set of distractors that you'd like to display in a controlled manner from 4 categories. They should not repeat (no distractor after distractor) and should be inserted into the image stream with even numbers of them from 4 categories.

Am I understanding this right?

The sample array function just picks a random image from the distractors list and ignores the category of the distractor. Is that the problem?

**Reply by ellamoeck** [reply]
Yes, so there's a target in every stream and they make their response to the target at the end of the stream. They don't have to respond to the fillers or distractors. 

The distractor appears either 2 or 8 items before the target, so in that respect the position of the target in the stream does matter. But you're correct in that the number of times each filler and target image appears is not important (and they are sampled from a set of 200+ images). In terms of placement in the stream, the distractors are already in their correct place, but they are randomly sampled from the two image categories (with 20 per category, or 10 per category if we are looking at the four categories) meaning that different images from the same category can appear more than twice in the experiment while others don't appear at all. Ideally, I would like each distractor image to appear twice. The sample array function isn't ignoring the category of the distractor (in that with of 80 distractor trials, 40 come from the reward and 40 from the neutral category), but it does ignore whether or not specific images have been shown previously. And because it's using two categories not all four, this issue is magnified (though even with all four categories programmed, I would still like to ensure that each distractor image is shown an equal number of times). Hopefully I'm making sense!

**Reply by vijaymarupudi** [reply]
In order to solve this problem ( I think), you need to maintain an image stack instead of picking a random item in the list of distractor images. Once you have an array of distractorImages, would need to pop an item off the image stack once it has been inserted into a stimulus array.

```javascript
function makeDistractorStack(providedDistractorImages) {
  const internalDistractorImagesCopy = providedDistractorImages.map(x => x)
  return function() {
    return internalDistractorImagesCopy.pop();
  }
}

distractorImageStack = makeDistractorStack(distractorImageArray)

// Anytime you need to insert a distractor, use:

[filler, filler, target, distractorImageStack(), target]  //... to output the image url

```

**Reply by ellamoeck** [reply]
Thanks for this suggestion. I've made those changes and the experiment runs, but when it gets to a distractor trial the screen is blank. I am not sure if I have entered the script correctly though - when you say "providedDistractorImages" do I need to use the name of the distractor images or do I just write "providedDistractorImages"? 

I've also put the distractorImagestack in like so (and not sure if this is correct):

```javascript
var runRSVPtrial = {
    type: 'rsvp-sequence',
    stimuli: function() {

        shuffleArray(standardFilenames);

        stimArray = [];
        for (ii = 0; ii < itemsPerStream + 2; ii++) {       // Extra 2 elements in array hold distractor and target stimuli; these will be moved into position later
            stimArray[ii] = standardFilenames[ii];
        };

        trialDistractType = distractType[trialSelector[trialNumInBlock]];

        if (trialDistractType == 1) {
            distractID = distractorImageStack(rewFilenames);
        } else if (trialDistractType == 2) {
            distractID = distractorImageStack(neutFilenames);
        } else if  (trialDistractType == 3) {
            distractID = standardFilenames[itemsPerStream];
        }

        stimArray[itemsPerStream] = distractID;

        trialTargetType = Math.floor(Math.random() * 2);
        if (trialTargetType == 0) {
            targetID = sampleArray(target0Filenames);
        } else {
            targetID = sampleArray(target1Filenames);
        };

        stimArray[itemsPerStream + 1] = targetID;

        return stimArray
    },
```

**Reply by vijaymarupudi** [reply]
You'll have to make a 'stack' using the function on the top. You want to provide that function the list of items you want to exclusively add into your stimulus list (i.e. the distractor image array probably). Then everytime you need a unique item from the 'stack', you call it like a function and it will returns the item (and removes it from the stack). You probably want to shuffle the array before providing it to the 'stack'.

Looking at this code, you might want to make two stacks, one for the rewFilenames and one for neutFilenames. And where you're currently calling the stack creator function, you should call the stack.

If you want the 'stack' to be unique experiment wide, make the 'stack' outside of the stimulus function. If you want it unique within a stimulus, them make it inside.

You should call makeDistractorStack with the list of images you want to make a unique stack of, like rewFilenames.

**Reply by ellamoeck** [reply]
That's fixed the issue! Thank you so much. One more quick question - it worked perfectly when I had each 40 distractor trials (so each distractor image showing once). But I've just increased to 80 trials and on the second round through, the distractor images are no longer showing. Is there a way of looping it so that it will get back and start calling the stack again?

**Reply by vijaymarupudi** [reply]
Make a stack that's twice as big (so initiate it with an array with 2x the times using `Array.concat()`). Or make 2 stacks.

**Reply by ellamoeck** [reply]
Great! Many thanks for your time and patience in providing additional explanations, I really appreciate it.

**Reply by vijaymarupudi** [reply]
You're welcome!
